song_scrounger/document_parser.py

```diff
@@ -1,49 +1,128 @@
-        tokens (list): strings found between quotes.
-
-    Notes:
-        - Ignores trailing quote if quotes are unbalanced
-        - Skips empty tokens
-    """
-    if len(text) == 0:
-        return []
-
-    tokens = []
-    while len(text) > 0:
-        quoted_token_indices = _find_first_two_quotes(text)
-        if quoted_token_indices is None:
-            break
-
-        opening_quote_index, closing_quote_index = quoted_token_indices
-        if opening_quote_index < closing_quote_index:
-            tokens.append(text[opening_quote_index+1:closing_quote_index])
-
-        text = "" if closing_quote_index+1 == len(text) else text[closing_quote_index+1:]
-    return tokens
-
-def _find_first_two_quotes(text):
-    """Finds indices of first two quotation marks.
-
-    e.g. 'A "quote"' => (2,8)
-    e.g. 'No quote' => None
-    e.g. 'Not "balanced' => None
-
-    Params:
-        text (str): e.g. 'A "quote"'.
-    Returns:
-        ((int,int)): indices of first two quotes in given text. None if absent or unbalanced.
-    """
-    if len(text) <= 1:
-        return None
-
-    opening_quote_index = text.find("\"")
-    if opening_quote_index != -1:
-        closing_quote_index = text.find("\"", opening_quote_index+1)
-        if closing_quote_index != -1:
-            return opening_quote_index, closing_quote_index
-    return None
\ No newline at end of file
+from .spotify_client import SpotifyClient
+from .models.song import Song
+
+
+class DocumentParser():
+    def __init__(self, spotify_client):
+        self.spotify_client = spotify_client
+
+    async def find_songs(self, text):
+        """Parses given text for songs, matching with artists if mentioned.
+
+        Each song is searched on Spotify. The artists in the search results
+        are searched for in the text as well. Any matches are used for
+        song disambiguation.
+
+        Params:
+            text (str): multi-paragraph page containing song names
+                and perhaps some of their artists.
+
+        Returns:
+            (dict): key (str) is song name; val ([str]) is spotify URIs
+                of matching songs, empty if no matching artist mentioned.
+        """
+        results = {}
+        paragraphs = self._get_paragraphs(text)
+        for paragraph in paragraphs:
+            song_names = self.find_quoted_tokens(paragraph)
+            for song_name in song_names:
+                songs = await self.search_spotify(song_name)
+                songs_with_mentioned_artists = self.filter_by_mentioned_artist(songs, text)
+                if len(songs_with_mentioned_artists) > 0:
+                    results[song_name] = [song.spotify_uri for song in songs_with_mentioned_artists]
+                else:
+                    results[song_name] = [song.spotify_uri for song in songs]
+        return results
+
+    def filter_by_mentioned_artist(self, songs, text):
+        songs_whose_artists_are_mentioned = set()
+        for song in songs:
```

Comment:
Cool thing I learned when playing around with list comprehension just now: set comprehension
`return { song if self.is_mentioned(artist, text) else None for song in songs for artist in song.artist }`

No reason to do this with stateful stuff but you can have this anywhere in place of loops which just build a list (slightly more performant too I believe).